packages/pubsub/src/connection-pool.js

```diff
@@ -0,0 +1,416 @@
+ * @param {object=} options - Pool options.
+ * @param {number} options.maxConnections - Number of connections to create.
+ *     Default: 5.
+ * @param {number} options.ackDeadline - The ack deadline to send when
+ *     creating a connection.
+ */
+function ConnectionPool(subscription) {
+  this.subscription = subscription;
+  this.connections = new Map();
+
+  this.isPaused = false;
+  this.isOpen = false;
+
+  this.failedConnectionAttempts = 0;
+  this.noConnectionsTime = Date.now();
+
+  this.settings = {
+    maxConnections: subscription.maxConnections || 5,
+    ackDeadline: subscription.ackDeadline || 10000
+  };
+
+  this.queue = [];
+  events.EventEmitter.call(this);
+
+  // grpc related fields we need since we're bypassing gax
+  this.metadata_ = new grpc.Metadata();
+
+  this.metadata_.add('x-goog-api-client', [
+    'gl-node/' + process.versions.node,
+    'gccl/' + PKG.version,
+    'grpc/' + require('grpc/package.json').version
+  ].join(' '));
+
+  this.open();
+}
+
+util.inherits(ConnectionPool, events.EventEmitter);
+
+/**
+ * Acquires a connection from the pool. Optionally you can specify an id for a
+ * specific connection, but if it is no longer available it will return the
+ * first available connection.
+ *
+ * @param {string=} id - The id of the connection to retrieve.
+ * @param {function} callback - The callback function.
+ * @param {?error} callback.err - An error returned while acquiring a
+ *     connection.
+ * @param {stream} callback.connection - A duplex stream.
+ */
+ConnectionPool.prototype.acquire = function(id, callback) {
+  var self = this;
+
+  if (is.fn(id)) {
+    callback = id;
+    id = null;
+  }
+
+  if (!this.isOpen) {
+    callback(new Error('No connections available to make request.'));
+    return;
+  }
+
+  // it's possible that by the time a user acks the connection could have
+  // closed, so in that case we'll just return any connection
+  if (!this.connections.has(id)) {
+    id = getFirstConnectionId();
+  }
+
+  var connection = this.connections.get(id);
+
+  if (connection) {
+    callback(null, connection);
+    return;
+  }
+
+  this.once('connected', function(connection) {
+    callback(null, connection);
+  });
+
+  function getFirstConnectionId() {
+    return self.connections.keys().next().value;
+  }
+};
+
+/**
+ * Ends each connection in the pool and closes the pool, preventing new
+ * connections from being created.
+ *
+ * @param {function} callback - The callback function.
+ * @param {?error} callback.error - An error returned while closing the pool.
+ */
+ConnectionPool.prototype.close = function(callback) {
+  var self = this;
+  var connections = Array.from(this.connections.values());
+
+  callback = callback || common.util.noop;
+
+  this.isOpen = false;
+  self.connections.clear();
+
+  this.queue.forEach(clearTimeout);
+  this.queue.length = 0;
+
+  each(connections, function(connection, onEndCallback) {
+    connection.end(onEndCallback);
+  }, callback);
+};
```

Comment:
You're right, I haven't been as diligent in cleaning up properties in this class because we usually throw a pool instance away when closing, but I will update for this.